File: packages/pen-core/src/variables/semantic-palette.ts
import type { PenDocument, VariableDefinition } from '@zseven-w/pen-types';

/**
 * 14-variable semantic palette for theme-aware element tools.
 *
 * Inventory rationale lives in
 * `openpencil-docs/superpowers/notes/2026-04-22-dark-theme-defaults-audit.md`.
 * That doc surveys the 30 hex literals in v0 element builders and
 * clusters them into these 14 semantic tokens — enough coverage
 * for every theme-dependent surface, small enough to stay
 * memorable.
 *
 * Every variable ships with BOTH light and dark values keyed on
 * a single theme axis `Mode`. Callers who want a single-theme
 * document (no switching) can drop the `themes` field and the
 * resolver will pick the first (light) value.
 *
 * This module is PURELY declarative — it does NOT mutate
 * `createEmptyDocument()`'s default output. v0 element tools
 * keep emitting hex literals (v0 byte-parity contract). v1
 * tools will `resolveVariableRef('$color-surface', doc.variables,
 * doc.themes ? {Mode: 'Dark'} : undefined)` to pick the right
 * shade at build time.
 *
 * To apply the palette to an existing document:
 *
 *   ```ts
 *   const themed = applySemanticPalette(doc);
 *   ```
 *
 * To get the resolved hex values for a specific mode (useful for
 * tests and tools that need a concrete color without full
 * document context):
 *
 *   ```ts
 *   const light = getSemanticPaletteHex('Light');
 *   // { 'color-surface': '#FFFFFF', 'color-accent': '#2563EB', ... }
 *   ```
 */

export const SEMANTIC_PALETTE_THEME_AXIS = 'Mode' as const;
export const SEMANTIC_PALETTE_THEME_LIGHT = 'Light' as const;
export const SEMANTIC_PALETTE_THEME_DARK = 'Dark' as const;

export type SemanticPaletteMode =
  | typeof SEMANTIC_PALETTE_THEME_LIGHT
  | typeof SEMANTIC_PALETTE_THEME_DARK;

export interface SemanticPalette {
  /** The `Mode: ['Light', 'Dark']` theme axis. Merge into `doc.themes`. */
  themes: Record<string, string[]>;
  /** The 14 variable definitions. Merge into `doc.variables`. */
  variables: Record<string, VariableDefinition>;
}

/**
 * Entry types for the PALETTE constant.
 * - `LightDark`: theme-aware color token (light + dark hex pair)
 * - `SingleColor`: single-value color token without theme axis (e.g. chart palette)
 * - `SingleNumber`: numeric token without theme axis (typography, spacing, radius)
 */
type LightDarkEntry = { light: string; dark: string; description: string };
type SingleColorEntry = { single: string; description: string };
type SingleNumberEntry = { single: number; description: string };
type PaletteEntry = LightDarkEntry | SingleColorEntry | SingleNumberEntry;

function isLightDark(e: PaletteEntry): e is LightDarkEntry {
  return 'light' in e;
}

/**
 * Raw value table. Source of truth — both
 * `getSemanticPalette()` and `getSemanticPaletteHex()` derive from
 * this. Keeping it as a plain object (not a function) makes the
 * palette trivially scanable in code review + reduces the risk of
 * values getting out of sync inside a function body.
 */
const PALETTE: Record<string, PaletteEntry> = {
  'color-surface': {
    light: '#FFFFFF',
    dark: '#1E293B',
    description: 'Primary surface — card, modal, tooltip background',
  },
  'color-surface-2': {
    light: '#F1F5F9',
    dark: '#334155',
    description: 'Secondary surface — chip, input, hover background',
  },
  'color-surface-3': {
    light: '#F3F4F6',
    dark: '#475569',
    description: 'Tertiary surface — pressed, deeper-nested background',
  },
  'color-bg-deep': {
    light: '#F8FAFC',
    dark: '#0F172A',
    description: 'Page background, skeleton hosts',
  },
  'color-border': {
    light: '#E2E8F0',
    dark: '#334155',
    description: 'Dividers, input strokes, card outlines',
  },
  'color-border-strong': {
    light: '#CBD5E1',
    dark: '#475569',
    description: 'Dashed chart-placeholder border',
  },
  'color-text-primary': {
    light: '#0F172A',
    dark: '#F1F5F9',
    description: 'Headlines, active page numbers',
  },
  'color-text-body': {
    light: '#334155',
    dark: '#CBD5E1',
    description: 'Body paragraphs, nav labels',
  },
  'color-text-muted': {
    light: '#64748B',
    dark: '#94A3B8',
    description: 'Secondary text, timestamps, placeholders',
  },
  'color-text-subtle': {
    light: '#94A3B8',
    dark: '#64748B',
    description: 'Tertiary text, disabled states',
  },
  'color-accent': {
    light: '#2563EB',
    dark: '#60A5FA',
    description: 'Primary brand, active pill, focus ring',
  },
  'color-destructive': {
    light: '#EF4444',
    dark: '#F87171',
    description: 'Delete actions, error state, trending-down',
  },
  'color-success': {
    light: '#10B981',
    dark: '#34D399',
    description: 'Trending-up, "Online" status',
  },
  'color-scrim': {
    light: '#00000080',
    dark: '#00000099',
    description: 'Modal backdrop',
  },

  // ── Alert tokens (8 light/dark pairs) ────────────────────────────────────
  'color-info-bg': {
    light: '#DBEAFE',
    dark: '#1E3A8A',
    description: 'Info alert background',
  },
  'color-info-text': {
    light: '#1E40AF',
    dark: '#BFDBFE',
    description: 'Info alert text / icon',
  },
  'color-success-bg': {
    light: '#DCFCE7',
    dark: '#14532D',
    description: 'Success alert background',
  },
  'color-success-text': {
    light: '#166534',
    dark: '#BBF7D0',
    description: 'Success alert text / icon',
  },
  'color-warning-bg': {
    light: '#FEF3C7',
    dark: '#78350F',
    description: 'Warning alert background',
  },
  'color-warning-text': {
    light: '#92400E',
    dark: '#FDE68A',
    description: 'Warning alert text / icon',
  },
  'color-danger-bg': {
    light: '#FEE2E2',
    dark: '#7F1D1D',
    description: 'Danger / error alert background',
  },
  'color-danger-text': {
    light: '#991B1B',
    dark: '#FECACA',
    description: 'Danger / error alert text / icon',
  },

  // ── Chart palette (6 single-value colors, no theme axis) ─────────────────
  'color-chart-1': { single: '#3B82F6', description: 'Chart series 1 — blue' },
  'color-chart-2': { single: '#8B5CF6', description: 'Chart series 2 — purple' },
  'color-chart-3': { single: '#EC4899', description: 'Chart series 3 — pink' },
  'color-chart-4': { single: '#14B8A6', description: 'Chart series 4 — teal' },
  'color-chart-5': { single: '#F59E0B', description: 'Chart series 5 — amber' },
  'color-chart-6': { single: '#F97316', description: 'Chart series 6 — orange' },
};

export const SEMANTIC_PALETTE_NAMES = Object.keys(PALETTE) as readonly (keyof typeof PALETTE)[];

/** Return the full palette as a `{themes, variables}` pair. */
export function getSemanticPalette(): SemanticPalette {
  const variables: Record<string, VariableDefinition> = {};
  for (const [name, entry] of Object.entries(PALETTE)) {
    if (isLightDark(entry)) {
      // Theme-aware color — light + dark ThemedValue array
      variables[name] = {
        type: 'color',
        value: [
          {
            value: entry.light,
            theme: { [SEMANTIC_PALETTE_THEME_AXIS]: SEMANTIC_PALETTE_THEME_LIGHT },
          },
          {
            value: entry.dark,
            theme: { [SEMANTIC_PALETTE_THEME_AXIS]: SEMANTIC_PALETTE_THEME_DARK },
          },
        ],
      };
    } else {
      // Single-value entry — no theme axis
      const isNum = typeof entry.single === 'number';
      variables[name] = {
        type: isNum ? 'number' : 'color',
        value: entry.single,
      };
    }
  }
  return {
    themes: {
      [SEMANTIC_PALETTE_THEME_AXIS]: [SEMANTIC_PALETTE_THEME_LIGHT, SEMANTIC_PALETTE_THEME_DARK],
    },
    variables,
  };
}

/**
 * Return just the resolved string/hex values for one mode. For theme-aware
 * (light/dark) entries the mode is respected. For single-value entries the
 * same value is returned regardless of mode. Numeric tokens are omitted.
 */
export function getSemanticPaletteHex(
  mode: SemanticPaletteMode = SEMANTIC_PALETTE_THEME_LIGHT,
): Record<string, string> {
  const out: Record<string, string> = {};
  for (const [name, entry] of Object.entries(PALETTE)) {
    if (isLightDark(entry)) {
      out[name] = mode === SEMANTIC_PALETTE_THEME_DARK ? entry.dark : entry.light;
    } else if (typeof entry.single === 'string') {
      // Single-value color (chart palette) — same for both modes
      out[name] = entry.single;
    }
    // SingleNumberEntry: not a color hex → omit from this map
  }
  return out;
}

/**
 * Return a human-readable description of a palette variable, for
 * tooling UI (token picker) and documentation generation.
 */
export function getSemanticPaletteDescription(name: string): string | undefined {
  return PALETTE[name]?.description;
}

/**
 * Merge the semantic palette into a PenDocument's `variables` and
 * `themes` fields. Returns a new document; does NOT mutate the
 * input. Existing variables/themes with the same names WIN — the
 * palette is purely additive when there's a collision, so a
 * document that has already defined `color-accent` differently
 * keeps its version.
 */
export function applySemanticPalette(doc: PenDocument): PenDocument {
  const palette = getSemanticPalette();
  const mergedVariables = { ...palette.variables, ...(doc.variables ?? {}) };
  const mergedThemes: Record<string, string[]> = { ...(doc.themes ?? {}) };
  for (const [axis, modes] of Object.entries(palette.themes)) {
    if (!mergedThemes[axis]) {
      mergedThemes[axis] = [...modes];
    }
  }
  return {
    ...doc,
    variables: mergedVariables,
    themes: mergedThemes,
  };
}

/**
 * Return `true` iff the document's variables include all 14
 * palette names. Useful as a runtime check before a v1 tool emits
 * `$color-*` refs — if the doc lacks the palette, the tool should
 * fall back to light-theme hex literals OR call
 * `applySemanticPalette()` to seed it.
 */
export function hasSemanticPalette(doc: PenDocument): boolean {
  const vars = doc.variables ?? {};
  return SEMANTIC_PALETTE_NAMES.every((name) => vars[name] !== undefined);
}

File: crates/op-orchestrator/src/semantic_palette.rs
//! shadcn-vocabulary semantic palette — the B1 dictionary seeded into
//! every generated document (replaces the retired 56-token `color-*`
//! palette; the old names keep resolving via `op-editor-core`'s
//! legacy compat table so pre-B1 documents still render).
//!
//! Composition (74 tokens, `--`-prefixed, same vocabulary as the
//! bundled `design-systems.json` presets and `apply_design_system`):
//! - 35 themed colour tokens (shadcn core 19 + sidebar 8 + status 8)
//! - 1 themed overlay token (`--scrim`)
//! - 6 single chart tokens (`--chart-1..6`, shadcn `--chart-*` shape)
//! - 2 font tokens (`--font-primary/--font-secondary`) — string
//! - 30 numeric tokens: 18 typography + 2 tracking + 5 spacing
//!   (unchanged legacy numerics) + 5 radius (`--radius-none/xs/m/l/pill`)
//!
//! Theme-aware tokens carry BOTH values on the single `Mode: [Light,
//! Dark]` axis; single-value entries have no theme axis. The palette
//! is what `$--token` refs emitted by theme="system" element builders
//! resolve against (paint-time resolution lives in
//! `op-editor-ui::scene_vars::VariableTable`).

use jian_ops_schema::variable::{
    ThemedValue, VariableDefinition, VariableKind, VariableScalar, VariableValue,
};
use std::collections::BTreeMap;

/// The single theme axis the palette declares.
pub const THEME_AXIS: &str = "Mode";
pub const THEME_LIGHT: &str = "Light";
pub const THEME_DARK: &str = "Dark";

enum Entry {
    /// Theme-aware colour — (light hex, dark hex).
    LightDark(&'static str, &'static str),
    /// Single-value colour, no theme axis (chart palette).
    Color(&'static str),
    /// Single-value string, no theme axis (fonts).
    Str(&'static str),
    /// Numeric token, no theme axis (typography / spacing / radius).
    Num(f64),
}

/// Light/dark pair, kept one line per entry so the table reads as a
/// dictionary.
macro_rules! ld {
    ($light:expr, $dark:expr) => {
        Entry::LightDark($light, $dark)
    };
}

/// Raw value table — source of truth. Values carry the slate identity
/// the retired `color-*` palette had, re-keyed to shadcn roles
/// (`--primary` is the old brand accent, `--muted` the old
/// `--muted`, `--accent` the old hover surface …).
const PALETTE: &[(&str, Entry)] = &[
    // ── shadcn core ───────────────────────────────────────────────
    ("--background", ld!("#F8FAFC", "#0F172A")),
    ("--foreground", ld!("#0F172A", "#F1F5F9")),
    ("--card", ld!("#FFFFFF", "#1E293B")),
    ("--card-foreground", ld!("#0F172A", "#F1F5F9")),
    ("--popover", ld!("#FFFFFF", "#1E293B")),
    ("--popover-foreground", ld!("#0F172A", "#F1F5F9")),
    ("--primary", ld!("#2563EB", "#60A5FA")),
    ("--primary-foreground", ld!("#FFFFFF", "#0F172A")),
    ("--secondary", ld!("#F1F5F9", "#334155")),
    ("--secondary-foreground", ld!("#0F172A", "#F1F5F9")),
    ("--muted", ld!("#F1F5F9", "#334155")),
    ("--muted-foreground", ld!("#64748B", "#94A3B8")),
    ("--accent", ld!("#F3F4F6", "#475569")),
    ("--accent-foreground", ld!("#0F172A", "#F1F5F9")),
    ("--destructive", ld!("#EF4444", "#F87171")),
    ("--destructive-foreground", ld!("#FFFFFF", "#0F172A")),
    ("--border", ld!("#E2E8F0", "#334155")),
    ("--input", ld!("#E2E8F0", "#334155")),
    ("--ring", ld!("#2563EB", "#60A5FA")),
    // ── sidebar family ────────────────────────────────────────────
    ("--sidebar", ld!("#FFFFFF", "#1E293B")),
    ("--sidebar-foreground", ld!("#0F172A", "#F1F5F9")),
    ("--sidebar-primary", ld!("#2563EB", "#60A5FA")),
    ("--sidebar-primary-foreground", ld!("#FFFFFF", "#0F172A")),
    ("--sidebar-accent", ld!("#F1F5F9", "#334155")),
    ("--sidebar-accent-foreground", ld!("#0F172A", "#F1F5F9")),
    ("--sidebar-border", ld!("#E2E8F0", "#334155")),
    ("--sidebar-ring", ld!("#2563EB", "#60A5FA")),
    // ── status colours ────────────────────────────────────────────
    ("--color-success", ld!("#10B981", "#34D399")),
    ("--color-success-foreground", ld!("#FFFFFF", "#0F172A")),
    ("--color-warning", ld!("#F59E0B", "#FBBF24")),
    ("--color-warning-foreground", ld!("#FFFFFF", "#0F172A")),
    ("--color-error", ld!("#EF4444", "#F87171")),
    ("--color-error-foreground", ld!("#FFFFFF", "#0F172A")),
    ("--color-info", ld!("#3B82F6", "#60A5FA")),
    ("--color-info-foreground", ld!("#FFFFFF", "#0F172A")),
    // ── overlay ───────────────────────────────────────────────────
    ("--scrim", ld!("#00000080", "#00000099")),
    // ── Typography: size + weight + line-height × 6 roles ─────────
    ("type-display-size", Entry::Num(64.0)),
    ("type-display-weight", Entry::Num(700.0)),
    ("type-display-line-height", Entry::Num(1.0)),
    ("type-h1-size", Entry::Num(24.0)),
    ("type-h1-weight", Entry::Num(600.0)),
    ("type-h1-line-height", Entry::Num(1.2)),
    ("type-h2-size", Entry::Num(20.0)),
    ("type-h2-weight", Entry::Num(600.0)),
    ("type-h2-line-height", Entry::Num(1.25)),
    ("type-h3-size", Entry::Num(16.0)),
    ("type-h3-weight", Entry::Num(600.0)),
    ("type-h3-line-height", Entry::Num(1.3)),
    ("type-body-size", Entry::Num(14.0)),
    ("type-body-weight", Entry::Num(400.0)),
    ("type-body-line-height", Entry::Num(1.5)),
    ("type-caption-size", Entry::Num(12.0)),
    ("type-caption-weight", Entry::Num(400.0)),
    ("type-caption-line-height", Entry::Num(1.4)),
    // ── Border radius ─────────────────────────────────────────────
    ("--radius-none", Entry::Num(0.0)),
    ("--radius-xs", Entry::Num(4.0)),
    ("--radius-m", Entry::Num(8.0)),
    ("--radius-l", Entry::Num(12.0)),
    ("--radius-pill", Entry::Num(999.0)),
    // ── Spacing scale ─────────────────────────────────────────────
    ("spacing-1", Entry::Num(4.0)),
    ("spacing-2", Entry::Num(8.0)),
    ("spacing-3", Entry::Num(12.0)),
    ("spacing-4", Entry::Num(16.0)),
    ("spacing-5", Entry::Num(24.0)),
    // ── Sparse letterSpacing ──────────────────────────────────────
    ("type-display-letter-spacing", Entry::Num(-0.5)),
    ("type-uppercase-label-letter-spacing", Entry::Num(1.5)),
    // ── Fonts ─────────────────────────────────────────────────────
    ("--font-primary", Entry::Str("Inter")),
    ("--font-secondary", Entry::Str("Space Grotesk")),
    // ── Chart palette ─────────────────────────────────────────────
    ("--chart-1", Entry::Color("#3B82F6")),
    ("--chart-2", Entry::Color("#8B5CF6")),
    ("--chart-3", Entry::Color("#EC4899")),
    ("--chart-4", Entry::Color("#14B8A6")),
    ("--chart-5", Entry::Color("#F59E0B")),
    ("--chart-6", Entry::Color("#F97316")),
];

/// Every palette token name, table order.
pub fn palette_names() -> Vec<&'static str> {
    PALETTE.iter().map(|(name, _)| *name).collect()
}

fn themed(mode: &str, hex: &str) -> ThemedValue {
    ThemedValue {
        value: VariableScalar::Str(hex.to_string()),
        theme: Some(BTreeMap::from([(THEME_AXIS.to_string(), mode.to_string())])),
    }
}

/// One token's `VariableDefinition`; `None` for unknown names.
pub fn palette_variable(name: &str) -> Option<VariableDefinition> {
    let entry = PALETTE.iter().find(|(n, _)| *n == name).map(|(_, e)| e)?;
    Some(match entry {
        Entry::LightDark(light, dark) => VariableDefinition {
            kind: VariableKind::Color,
            value: VariableValue::Themed(vec![
                themed(THEME_LIGHT, light),
                themed(THEME_DARK, dark),
            ]),
        },
        Entry::Color(hex) => VariableDefinition {
            kind: VariableKind::Color,
            value: VariableValue::Scalar(VariableScalar::Str(hex.to_string())),
        },
        Entry::Str(s) => VariableDefinition {
            kind: VariableKind::String,
            value: VariableValue::Scalar(VariableScalar::Str(s.to_string())),
        },
        Entry::Num(n) => VariableDefinition {
            kind: VariableKind::Number,
            value: VariableValue::Scalar(VariableScalar::Num(*n)),
        },
    })
}

/// The full palette as a variables map (for `MergeThemePreset`).
pub fn palette_variables() -> BTreeMap<String, VariableDefinition> {
    PALETTE
        .iter()
        .map(|(name, _)| {
            (
                name.to_string(),
                palette_variable(name).expect("table name"),
            )
        })
        .collect()
}

/// The `Mode: [Light, Dark]` theme axis declaration.
pub fn palette_themes() -> BTreeMap<String, Vec<String>> {
    BTreeMap::from([(
        THEME_AXIS.to_string(),
        vec![THEME_LIGHT.to_string(), THEME_DARK.to_string()],
    )])
}

#[cfg(test)]
mod tests {
    use super::*;

    /// 74 tokens: 36 themed colours + 6 chart singles + 2 font
    /// strings + 30 numerics — the B1 shadcn dictionary.
    #[test]
    fn palette_inventory_matches_b1_dictionary() {
        let vars = palette_variables();
        assert_eq!(vars.len(), 74);
        let themed = vars
            .values()
            .filter(|v| matches!(v.value, VariableValue::Themed(_)))
            .count();
        let numbers = vars
            .values()
            .filter(|v| v.kind == VariableKind::Number)
            .count();
        assert_eq!(themed, 36, "19 shadcn core + 8 sidebar + 8 status + scrim");
        assert_eq!(numbers, 30, "18 type + 2 tracking + 5 spacing + 5 radius");
        assert_eq!(vars.len() - themed - numbers, 8, "6 chart + 2 font singles");
    }

    /// Spot-check values stay stable with the retired palette's
    /// identity: the brand accent moved to `--primary`, the page
    /// ground to `--background`.
    #[test]
    fn spot_values_verbatim() {
        let primary = palette_variable("--primary").unwrap();
        let VariableValue::Themed(pair) = &primary.value else {
            panic!("primary must be themed");
        };
        assert_eq!(pair[0].value, VariableScalar::Str("#2563EB".into()));
        assert_eq!(pair[1].value, VariableScalar::Str("#60A5FA".into()));
        assert_eq!(
            pair[0].theme.as_ref().unwrap().get(THEME_AXIS),
            Some(&THEME_LIGHT.to_string())
        );

        let body = palette_variable("type-body-size").unwrap();
        assert_eq!(body.value, VariableValue::Scalar(VariableScalar::Num(14.0)));

        let font = palette_variable("--font-primary").unwrap();
        assert_eq!(
            font.value,
            VariableValue::Scalar(VariableScalar::Str("Inter".into()))
        );

        let pill = palette_variable("--radius-pill").unwrap();
        assert_eq!(
            pill.value,
            VariableValue::Scalar(VariableScalar::Num(999.0))
        );
    }

    #[test]
    fn themes_declare_mode_axis() {
        let themes = palette_themes();
        assert_eq!(
            themes.get(THEME_AXIS),
            Some(&vec!["Light".to_string(), "Dark".to_string()])
        );
    }
}
